src/azure-cli-core/azure/cli/core/extension.py

```diff
@@ -0,0 +1,150 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+import os
+import traceback
+import json
+
+from wheel.install import WHEEL_INFO_RE
+
+import azure.cli.core.azlogging as azlogging
+
+from azure.cli.core._config import GLOBAL_CONFIG_DIR
+
+
+_CUSTOM_EXT_DIR = os.environ.get('AZURE_EXTENSION_DIR')
+EXTENSIONS_DIR = os.path.expanduser(_CUSTOM_EXT_DIR) if _CUSTOM_EXT_DIR \
+                    else os.path.join(GLOBAL_CONFIG_DIR, 'cliextensions')
+EXTENSIONS_MOD_PREFIX = 'azext_'
+
+logger = azlogging.get_az_logger(__name__)
+
+logger.debug("Extension directory: '%s'", EXTENSIONS_DIR)
+
+
+class ExtensionNotInstalledException(Exception):
+    def __init__(self, extension_name):
+        super(ExtensionNotInstalledException, self).__init__(extension_name)
+        self.extension_name = extension_name
+
+    def __str__(self):
+        return "The extension {} is not installed.".format(self.extension_name)
+
+
+class Extension(object):
+
+    def __init__(self, name, ext_type):
+        self.name = name
+        self.ext_type = ext_type
+        self._version = None
+        self._metadata = None
+
+    @property
+    def version(self):
+        """
+        Lazy load version.
+        Returns the version as a string or None if not available.
+        """
+        try:
+            self._version = self._version or self.get_version()
+        except Exception:  # pylint: disable=broad-except
+            logger.debug("Unable to get extension version: %s", traceback.format_exc())
```

Comment:
Yes so they *should* have versions but I'm being conservative here as don't want the CLI to crash just because an extension doesn't have a version. This is why I'm also logging to `debug` so the error message is not shown to the average user in the rare case it does happen.